### PR TITLE
compiler-rt: Support thumb versions older than armv6

### LINKED ISSUE #2738: build-exe for thumbv4t-freestanding fails with "instruction variant requires ARMv6 or later"
I'd like to use Zig to write software for the Game Boy Advance. The GBA uses the ARMv4T instruction set, and it accesses the cartridge ROM using a 16-bit bus, so I'm thinking the best target to use is `thumbv4t-freestanding`. But attempting to run `zig build-exe` with that target results in the following error:

```
$ echo 'pub fn main() void {}' > trivial.zig
$ zig build-exe trivial.zig -target thumbv4t-freestanding
<inline asm>:1:3: error: instruction variant requires ARMv6 or later
         mov     r3, r1
         ^
<inline asm>:2:2: error: instruction variant requires ARMv6 or later
 mov     r1, r2
 ^
<inline asm>:3:2: error: instruction variant requires ARMv6 or later
 mov     r2, r3
 ^
LLVM ERROR: Error parsing inline asm

$
```

I suspect this message is referring to [the definition of `__aeabi_memset` in `compiler_rt.zig`](https://github.com/ziglang/zig/blob/24400d5882aa2b83a4aed3e4c1503d0393e1c541/std/special/compiler_rt.zig#L510). That function (along with `__aeabi_memclr`) has `mov` instructions where both the source and destination are low registers. [A "superseded" reference guide](http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.dui0204j/Cihcdbca.html) reveals that this is invalid prior to ARMv6:

> `MOV Rd, Rm`
>
>    In architectures before ARMv6, either Rd or Rm, or both, must be a Hi register. In ARMv6 and above, this restriction does not apply.

### PR DESCRIPTION
**Pull request by cartr**
This PR fixes #2738 by adding versions of `__aeabi_memset` and `__aeabi_memclr` which use instructions supported on thumbv4. It replaces the register-register movs in `__aeabi_memset` with a XOR swap, and it replaces the mov in `__aeabi_memclr` with an `adds` instruction.

I'd appreciate it if someone more familiar with the ARM ABI/assembly language could review this -- as it stands, I've confirmed a trivial program now compiles without errors, but I'm not sure how to check if the functions I changed actually behave correctly in all situations.

### CODE REVIEW

File: std/special/compiler_rt.zig [hunk 1]
```diff
@@ -471,6 +471,28 @@ test "usesThumb1" {
     //etc.
 }
 
+const use_thumb_1_pre_armv6 = usesThumb1PreArmv6(builtin.arch);
+
+fn usesThumb1PreArmv6(arch: builtin.Arch) bool {
+    return switch (arch) {
+        .thumb => switch (arch.thumb) {
```
Comment (andrewrk) [inline std/special/compiler_rt.zig]: Slightly more idiomatic:
```suggestion
        .thumb => |sub_arch |switch (sub_arch) {
```

File: std/special/compiler_rt.zig [hunk 2]
```diff
@@ -471,6 +471,28 @@ test "usesThumb1" {
...
+            .v5,
+            .v5te,
+            .v4t,
+            => true,
+            else => false,
+        },
+        .thumbeb => switch (arch.thumbeb) {
```
Comment (andrewrk) [inline std/special/compiler_rt.zig]:
```suggestion
        .thumbeb => |sub_arch| switch (sub_arch) {
```

**Comment by andrewrk** [comment]
@cartr looks like you're the pioneer of older thumb architectures in Zig!

File: std/special/compiler_rt.zig [hunk 3]
```diff
@@ -505,7 +527,16 @@ nakedcc fn __aeabi_memmove() noreturn {
 
 nakedcc fn __aeabi_memset() noreturn {
     @setRuntimeSafety(false);
-    if (use_thumb_1) {
+    if (use_thumb_1_pre_armv6) {
+        asm volatile (
+            \\ eors    r1, r2
```
Comment (SamTebbs33) [inline std/special/compiler_rt.zig]: I've only worked on recent versions of arm and aarch64 so I'm not fully clued up on the intricacies of thumb, but can you use `eor` instead? The only difference AFAIK is that it doesn't update the condition flags. Same goes for `adds` -> `add` and `movs` -> `mov` in `memclr`.

**Comment by andrewrk** [comment]
Ooh, there's an even better way to express this. Check this out:

```zig
const use_thumb_1_pre_armv6 = switch (builtin.arch) {
    .thumb, .thumbeb => |sub_arch| switch (sub_arch) {
        .v5, .v5te, .v4t => true,
        else => false,
    },
    else => false,
};
```

Since `thumb` and `thumbeb` have the same payload type, you can capture the `sub_arch` for both.

**Comment by cartr** [comment]
Thanks for the feedback! @andrewrk, I've updated my PR to use the syntax you suggested, but the "even better way" from your last comment results in an error on my machine:

```
/home/carter/repos/zig/std/special/compiler_rt.zig:476:9: error: enum '@TagType(builtin.Arch)' has no field named 'v5'
        .v5, .v5te, .v4t => true,
        ^
/home/carter/.local/share/zig/stage1/builtin/QBjTG78Bsf62AJDqbRcdXHTQssc8HBZyajT-kiWnsru68ajowOmdwr7OMuz3SHyu/builtin.zig:47:18: note: '@TagType(builtin.Arch)' declared here
pub const Arch = union(enum) {
```

It looks like it's capturing the tag instead of the value. (I'm not super well-versed in Zig yet -- is this a bug?)

File: std/special/compiler_rt.zig (hunk 3 above)
Comment (cartr) [reply std/special/compiler_rt.zig]: I don't think the non-flag-updating versions of these instructions are available in thumbv4t -- [the GBATEK document I'm using as a reference](https://problemkaputt.de/gbatek.htm#thumbinstructionsummary) just calls this instruction `EOR` and says it always updates the condition flags.

**Comment by andrewrk** [comment]
> It looks like it's capturing the tag instead of the value. (I'm not super well-versed in Zig yet -- is this a bug?)

Ah my mistake. I tested the snippet before making that comment but did not check the type of the capture value.

It's either a bug or a missing feature that I mistakenly thought I had already implemented.

**Comment by andrewrk** [comment]
Thanks for the patch!

File: std/special/compiler_rt.zig (hunk 3 above)
Comment (SamTebbs33) [reply std/special/compiler_rt.zig]: Yes you're right, I just got an ARMv6t ISA reference card from someone at work and there is no `EOR`, only `EORS`.

**Pull request merged**